### ISSUE TITLE
Increase reconnect delay

### ISSUE THREAD

**Issue by Ninir** [elasticsearch/elasticsearch-river-twitter#91]
Hello guys!

I am working on a project which handles a long-running process of retrieving tweets using the river (let's say long-running = 1 year). Sometimes, the stream is failing / closing and restarting. This is quite annoying BUT the reconnect stuff is working like a charm! :+1: 

However, the delay to reconnect to the Stream API after a failure is too quick, after what it keeps getting *420 errors*.

#### StackTrace (for a 2seconds window)
```
[2015-01-27 00:23:53,085][WARN ][river.twitter            ] There was failures while executing bulk
[2015-01-27 00:24:13,084][WARN ][river.twitter            ] There was failures while executing bulk
[2015-01-27 00:24:43,085][WARN ][river.twitter            ] There was failures while executing bulk
[2015-01-27 00:26:13,085][WARN ][river.twitter            ] There was failures while executing bulk
[2015-01-27 00:28:06,532][INFO ][twitter4j.TwitterStreamImpl] Stream closed.
[2015-01-27 00:28:06,532][WARN ][river.twitter            ] [Taj Nital] [twitter][twitter-river] stream failure, restarting stream...
Stream closed.
Relevant discussions can be found on the Internet at:
        http://www.google.co.jp/search?q=a8fd061d or
        http://www.google.co.jp/search?q=00076b7e
TwitterException{exceptionCode=[a8fd061d-00076b7e a8fd061d-00076b0b], statusCode=-1, message=null, code=-1, retryAfter=-1, rateLimitStatus=null, version=4.0.2}
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:190)
        at twitter4j.StatusStreamImpl.next(StatusStreamImpl.java:56)
        at twitter4j.TwitterStreamImpl$TwitterStreamConsumer.run(TwitterStreamImpl.java:504)
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.read(SocketInputStream.java:152)
        at java.net.SocketInputStream.read(SocketInputStream.java:122)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:442)
        at sun.security.ssl.InputRecord.read(InputRecord.java:480)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:927)
        at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:884)
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:102)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
        at sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:552)
        at sun.net.www.http.ChunkedInputStream.readAhead(ChunkedInputStream.java:609)
        at sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:696)
        at java.io.FilterInputStream.read(FilterInputStream.java:133)
        at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3053)
        at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:238)
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
        at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:116)
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.fill(BufferedReader.java:154)
        at java.io.BufferedReader.readLine(BufferedReader.java:317)
        at java.io.BufferedReader.readLine(BufferedReader.java:382)
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:75)
        ... 2 more
[2015-01-27 00:28:06,533][INFO ][twitter4j.TwitterStreamImpl] Stream closed.
[2015-01-27 00:28:06,533][WARN ][river.twitter            ] [Taj Nital] [twitter][twitter-river] stream failure, restarting stream...
[2015-01-27 00:28:06,534][INFO ][twitter4j.TwitterStreamImpl] Waiting for 250 milliseconds
[2015-01-27 00:28:06,784][INFO ][twitter4j.TwitterStreamImpl] Establishing connection.
[2015-01-27 00:28:06,869][INFO ][twitter4j.TwitterStreamImpl] Stream closed.
[2015-01-27 00:28:06,869][WARN ][river.twitter            ] [Taj Nital] [twitter][twitter-river] stream failure, restarting stream...
Stream closed.
Relevant discussions can be found on the Internet at:
        http://www.google.co.jp/search?q=a8fd061d or
        http://www.google.co.jp/search?q=00076b7e
TwitterException{exceptionCode=[a8fd061d-00076b7e a8fd061d-00076b0b], statusCode=-1, message=null, code=-1, retryAfter=-1, rateLimitStatus=null, version=4.0.2}
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:190)
        at twitter4j.StatusStreamImpl.next(StatusStreamImpl.java:56)
        at twitter4j.TwitterStreamImpl$TwitterStreamConsumer.run(TwitterStreamImpl.java:504)
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.read(SocketInputStream.java:152)
        at java.net.SocketInputStream.read(SocketInputStream.java:122)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:442)
        at sun.security.ssl.InputRecord.read(InputRecord.java:480)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:927)
        at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:884)
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:102)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
        at sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:552)
        at sun.net.www.http.ChunkedInputStream.readAhead(ChunkedInputStream.java:609)
        at sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:696)
        at java.io.FilterInputStream.read(FilterInputStream.java:133)
        at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3053)
        at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:238)
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
        at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:116)
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.fill(BufferedReader.java:154)
        at java.io.BufferedReader.readLine(BufferedReader.java:317)
        at java.io.BufferedReader.readLine(BufferedReader.java:382)
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:75)
        ... 2 more
[2015-01-27 00:28:06,870][INFO ][twitter4j.TwitterStreamImpl] Stream closed.
[2015-01-27 00:28:06,870][WARN ][river.twitter            ] [Taj Nital] [twitter][twitter-river] stream failure, restarting stream...
[2015-01-27 00:28:06,870][INFO ][twitter4j.TwitterStreamImpl] Stream closed.
[2015-01-27 00:28:06,870][WARN ][river.twitter            ] [Taj Nital] [twitter][twitter-river] stream failure, restarting stream...
Stream closed.
Relevant discussions can be found on the Internet at:
        http://www.google.co.jp/search?q=a8fd061d or
        http://www.google.co.jp/search?q=00076b7e
TwitterException{exceptionCode=[a8fd061d-00076b7e a8fd061d-00076b0b], statusCode=-1, message=null, code=-1, retryAfter=-1, rateLimitStatus=null, version=4.0.2}
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:190)
        at twitter4j.StatusStreamImpl.next(StatusStreamImpl.java:56)
        at twitter4j.TwitterStreamImpl$TwitterStreamConsumer.run(TwitterStreamImpl.java:504)
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.read(SocketInputStream.java:152)
        at java.net.SocketInputStream.read(SocketInputStream.java:122)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:442)
        at sun.security.ssl.InputRecord.read(InputRecord.java:480)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:927)
        at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:884)
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:102)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
        at sun.net.www.http.ChunkedInputStream.readAheadBlocking(ChunkedInputStream.java:552)
        at sun.net.www.http.ChunkedInputStream.readAhead(ChunkedInputStream.java:609)
        at sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:696)
        at java.io.FilterInputStream.read(FilterInputStream.java:133)
        at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3053)
        at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:238)
        at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
        at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:116)
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.fill(BufferedReader.java:154)
        at java.io.BufferedReader.readLine(BufferedReader.java:317)
        at java.io.BufferedReader.readLine(BufferedReader.java:382)
        at twitter4j.StatusStreamBase.handleNextElement(StatusStreamBase.java:75)
        ... 2 more
[2015-01-27 00:28:06,870][INFO ][twitter4j.TwitterStreamImpl] Waiting for 250 milliseconds
[2015-01-27 00:28:07,120][INFO ][twitter4j.TwitterStreamImpl] Establishing connection.
[2015-01-27 00:28:07,476][INFO ][twitter4j.TwitterStreamImpl] 420:Returned by the Search and Trends API when you are being rate limited (https://dev.twitter.com/docs/rate-limiting).
Returned by the Streaming API:
 Too many login attempts in a short period of time.
 Running too many copies of the same application authenticating with the same account name.
Exceeded connection limit for user
```

#### Idea: Increasing delay to reconnect
@dadoonet Do you think it would be possible to increase the time after which it reconnects?
or at least allowing to configure it?

I'm always forced to stop ES (I can't remove the river because it is dynamically updated so I would be forced to get it, and reput it after some time) for 20-30s, and then restart it.

As I'm quite a beginner to ES and rivers in general, you may have better tips! ;)

Thank your for your time!

**Comment by schmorgs** [comment]
Good timing for this post.  I've been seeing this too lately, and the fix for me is to delete the river, wait a couple of minutes, then recreate it.

But yes, if we could set a reconnect delay when rate exceeded "error 420" is seen, it might help here.

**Comment by dadoonet** [comment]
Very interesting indeed. Do you know that Logstash actually supports it already? 
Look at this line: https://github.com/logstash-plugins/logstash-input-twitter/blob/master/lib/logstash/inputs/twitter.rb#L115

**Comment by Ninir** [comment]
@dadoonet I do now! :) I'll take a look at it later, thank you !
do you think it would be possible to add something like this, just for the help it brings for now? :)

**Comment by schmorgs** [comment]
Hi @dadoonet, the problem with Logstash at the moment is it doesn't support proxy servers so we can't get to Twitter from within our  company.  I raised this as a bug on the logstash channel, but there's been no progress on that for a while so for the moment, it's a no-go.

**Comment by dadoonet** [comment]
@schmorgs Fair point. I opened an issue here: https://github.com/logstash-plugins/logstash-input-twitter/issues/7

**Comment by Ninir** [comment]
@dadoonet For now, is it about changing the value of `TimeValue.timeValueSeconds(10)`: https://github.com/elasticsearch/elasticsearch-river-twitter/blob/master/src/main/java/org/elasticsearch/river/twitter/TwitterRiver.java#L548

for a value defined in the conf?

**Comment by dadoonet** [comment]
@Ninir I agree. It's not a big change and can be seen as a bug fix somehow useful in the meantime. `retry_after` could be the name? 

Wanna come with a PR?

**Comment by dadoonet** [comment]
Closed with #92.
